lib/coll.gi

```diff
@@ -1682,496 +1682,230 @@ InstallMethod( ForAnyOp,
-            for val  in gen  do
-                vals[l+1] := val;
-                ListXHelp( result, gens, i+1, vals, l+1 );
-            od;
-            Unbind( vals[l+1] );
-            return;
-        else
-            Error( "gens[",i+1,"] must be a collection, a list, a boolean, ",
-                   "or a function" );
-        fi;
-    od;
-    Add( result, CallFuncList( gens[i+1], vals ) );
+InstallGlobalFunction( ListX, function ( arg )
+    local f;
+    f := Remove(arg);
+    return FoldLeftX(arg,
+            function(acc, x)
+                Add(acc, CallFuncList(f, x));
+                return acc;
+            end, []);
 end );
 
-BIND_GLOBAL( "ListXHelp2", function ( result, gens, i, val1, val2 )
-    local   gen, vals, val3;
-    while i+1 < Length(gens)  do
-        gen := gens[i+1];
-        if IsFunction( gen )  then
-            gen := gen( val1, val2 );
-        fi;
-        if gen = true  then
-            i := i + 1;
-        elif gen = false  then
-            return;
-        elif IsListOrCollection( gen )  then
-            vals := [ val1, val2 ];
-            for val3  in gen  do
-                vals[3] := val3;
-                ListXHelp( result, gens, i+1, vals, 3 );
-            od;
-            Unbind( vals[3] );
-            return;
-        else
-            Error( "gens[",i+1,"] must be a collection, a list, a boolean, ",
-                   "or a function" );
-        fi;
-    od;
-    Add( result, gens[i+1]( val1, val2 ) );
-end );
 
-BIND_GLOBAL( "ListXHelp1", function ( result, gens, i, val1 )
-    local   gen, val2;
-    while i+1 < Length(gens)  do
-        gen := gens[i+1];
-        if IsFunction( gen )  then
-            gen := gen( val1 );
-        fi;
-        if gen = true  then
-            i := i + 1;
-        elif gen = false  then
-            return;
-        elif IsListOrCollection( gen )  then
-            for val2  in gen  do
-                ListXHelp2( result, gens, i+1, val1, val2 );
-            od;
-            return;
+#############################################################################
+##
+#F  FoldLeft( <coll>, <func> )
+#F  FoldLeft( <coll>, <func>, <init> )
+##
+InstallGlobalFunction( FoldLeft,
+    function( arg )
+    local tnum, C, func, result, i, l;
+    l := Length( arg );
+    if l < 2 or l > 3 or not IsFunction(arg[2]) then
+      Error( "usage: FoldLeft( <C>, <func>[, <init>] )" );
+    fi;
+    tnum:= TNUM_OBJ( arg[1] );
+    # handle built-in lists directly, to avoid method dispatch overhead
+    if FIRST_LIST_TNUM <= tnum and tnum <= LAST_LIST_TNUM then
+      C:= arg[1];
+      func:= arg[2];
+      if l = 2 then
+        if IsEmpty( C ) then
+          Error("folding an empty collection without initial value is not supported");
         else
-            Error( "gens[",i+1,"] must be a collection, a list, a boolean, ",
-                   "or a function" );
+          result:= C[1];
+          for i in [ 2 .. Length( C ) ] do
```

Comment:
Contrasting this with the code path used when `init` is given (which uses `for i in C do`) I think we get different behavior if `C` has holes. We probably don't want that difference, do we?